Change default for `unknown` to `RAISE`

lafrech opened the pull request:
Closes #851 

The change to the code is minimal but a lot of tests needed to be fixed.

Most tests are fixed by just passing `unknown=EXCLUDE` not out of slopiness but because they check that good keys are passed and bad keys are not, so raising wouldn't be practical there.

I removed the unknown keys where they were not relevant to the test.

I also modified `test_load_resets_error_fields` as I think it was broken and it is better this way.

lafrech committed:
Change default for `unknown` to `RAISE`

Closes #851

sloria commented on tests/test_deserialization.py:
```diff
@@ -1354,15 +1354,15 @@ def test_unknown_fields_deserialization(self):
             foo = fields.Integer()
 
-        data = MySchema().load({'foo': 3, 'bar': 5})
+        data = MySchema(unknown=EXCLUDE).load({'foo': 3, 'bar': 5})
```
I think this test was meant to test the default behavior when `unknown` isn't passed. `unknown=EXCLUDE` is tested a few lines down. So I think this should test that an error is raised.

lafrech replied:
Default is tested a few lines later.

Maybe this test could be split in several tests and/or parametrized, but I think it does the right thing. At least I didn't change it, I just switched the `EXCLUDE` and `RAISE` cases.

sloria replied:
Oh, I missed that. OK, I think this if fine as is.

sloria commented on tests/test_decorators.py:
```diff
@@ -550,7 +550,7 @@ def strict_fields(self, data, original_data):
                         raise ValidationError('Unknown field name', key)
 
-        schema = MySchema()
+        schema = MySchema(unknown=EXCLUDE)
```
This test is fine for now, but it'd be good to change it soon, now that there's a built-in way to handle unknown fields.

lafrech replied:
I'm not sure I understand.

Do you mean that it is good as a test but bad as an example, so we should change it to a more useful or at least more neutral example?

sloria replied:
Exactly: the `strict_fields` implementation is a bad example that should be replaced by a more neutral one.

lafrech replied:
Done.

sloria reviewed (commented):
Just a minor change suggested. Otherwise, this looks good to go.

lafrech committed:
Rewrite test_allow_arbitrary_field_names_in_error

The former implementation of this test illustrates a use case involving
a schema validator that is now useless since its function (raising an
error on unknown keys) is a feature of Marshmallow. This rewrite makes
the test simpler and more neutral.

sloria commented:
Thanks @lafrech . Would you like to update the changelog and the upgrading guide?

If you don't have time right now, I'll try to find some time during the weekend.

lafrech commented:
Changelog + upgrading guide updated in dev branch. Feel free to reword.

sloria commented:
Thanks!
